### PR TITLE
Stop using `PolyTraitRef` for closure/coroutine predicates already instantiated w placeholders

### PR DESCRIPTION
**Pull request by compiler-errors**
r? lcnr

### CODE REVIEW

**Comment by compiler-errors** [comment]
lmao

@rustbot author

File: compiler/rustc_trait_selection/src/traits/error_reporting/type_err_ctxt_ext.rs [hunk 1]
```diff
@@ -3394,7 +3394,12 @@ impl<'tcx> TypeErrCtxt<'_, 'tcx> {
             (obligation.cause.clone(), terr)
         };
         self.report_and_explain_type_error(
-            TypeTrace::poly_trait_refs(&cause, true, expected_trait_ref, found_trait_ref),
+            TypeTrace::poly_trait_refs(
```
Comment (lcnr) [inline compiler/rustc_trait_selection/src/traits/error_reporting/type_err_ctxt_ext.rs:3397]: add or use `TypeTrace::trait_refs`

File: compiler/rustc_trait_selection/src/traits/error_reporting/type_err_ctxt_ext.rs [hunk 2]
```diff
@@ -3487,15 +3489,15 @@ impl<'tcx> TypeErrCtxt<'_, 'tcx> {
         // traits manually, but don't make it more confusing when it does
         // happen.
         Ok(
-            if Some(expected_trait_ref.def_id()) != self.tcx.lang_items().coroutine_trait()
+            if Some(expected_trait_ref.def_id) != self.tcx.lang_items().coroutine_trait()
                 && not_tupled
             {
                 self.report_and_explain_type_error(
                     TypeTrace::poly_trait_refs(
```
Comment (lcnr) [inline compiler/rustc_trait_selection/src/traits/error_reporting/type_err_ctxt_ext.rs:3496]: same here

File: compiler/rustc_trait_selection/src/traits/select/confirmation.rs [hunk 3]
```diff
@@ -1079,8 +1079,8 @@ impl<'cx, 'tcx> SelectionContext<'cx, 'tcx> {
             })
             .map_err(|terr| {
                 SignatureMismatch(Box::new(SignatureMismatchData {
```
Comment (lcnr) [inline compiler/rustc_trait_selection/src/traits/select/confirmation.rs:1081]: is that sometimes used with an uninstantiated binder?

**Review by lcnr — COMMENTED** [review]
r=me after nits

File: compiler/rustc_trait_selection/src/traits/select/confirmation.rs (hunk 3 above)
Comment (compiler-errors) [reply compiler/rustc_trait_selection/src/traits/select/confirmation.rs:1081]: nope, it's always instantiated

Comment (lcnr) [reply compiler/rustc_trait_selection/src/traits/select/confirmation.rs:1081]: then we can remove the `Binder` from `SignatureMismatchData` :thinking:

Comment (compiler-errors) [reply compiler/rustc_trait_selection/src/traits/select/confirmation.rs:1081]: isn't that what this PR is doing? 🤔

Comment (lcnr) [reply compiler/rustc_trait_selection/src/traits/select/confirmation.rs:1081]: ?????????????????????????????????????????????? I might have been utterly confused

god knows what i was thinking. sorry

**Comment by lcnr** [comment]
sorry, apparently unable to read diffs

@bors r+ rollup

**Pull request merged**